framework/behaviors/AttributeTypecastBehavior.php

```diff
@@ -0,0 +1,329 @@
+     * in format: ownerClassName => attributeTypes
+     */
+    private static $autoDetectedAttributeTypes = [];
+
+
+    /**
+     * Clears internal static cache of auto detected [[attributeTypes]] values
+     * over all affected owner classes.
+     */
+    public static function clearAutoDetectedAttributeTypes()
+    {
+        self::$autoDetectedAttributeTypes = [];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function init()
+    {
+        parent::init();
+
+        if ($this->attributeTypes === null) {
+            $ownerClass = get_class($this->owner);
+            if (!isset(self::$autoDetectedAttributeTypes[$ownerClass])) {
+                self::$autoDetectedAttributeTypes[$ownerClass] = $this->detectAttributeTypes();
+            }
+            $this->attributeTypes = self::$autoDetectedAttributeTypes[$ownerClass];
+        }
+    }
+
+    /**
+     * Typecast owner attributes according to [[attributeTypes]].
+     * @param array $attributeNames list of attribute names that should be type-casted.
+     * If this parameter is empty, it means any attribute listed in the [[attributeTypes]]
+     * should be type-casted.
+     */
+    public function typecastAttributes($attributeNames = null)
+    {
+        $attributeTypes = [];
+
+        if ($attributeNames === null) {
+            $attributeTypes = $this->attributeTypes;
+        } else {
+            foreach ($attributeNames as $attribute) {
+                if (!isset($this->attributeTypes[$attribute])) {
+                    throw new InvalidParamException("There is no type mapping for '{$attribute}'.");
+                }
+                $attributeTypes[$attribute] = $this->attributeTypes[$attribute];
+            }
+        }
+
+        foreach ($attributeTypes as $attribute => $type) {
+            $this->owner->{$attribute} = $this->typecastValue($this->owner->{$attribute}, $type);
+        }
+    }
+
+    /**
+     * Casts the given value to the specified type.
+     * @param mixed $value value to be type-casted.
+     * @param string|callable $type type name or typecast callable.
+     * @return mixed typecast result.
+     */
+    protected function typecastValue($value, $type)
+    {
+        if ($this->skipOnNull && $value === null) {
+            return $value;
+        }
+
+        if (is_scalar($type)) {
+            if (is_object($value) && method_exists($value, '__toString')) {
+                $value = $value->__toString();
+            }
+
+            switch ($type) {
+                case self::TYPE_INTEGER:
+                    return (int)$value;
+                case self::TYPE_FLOAT:
+                    return (float)$value;
+                case self::TYPE_BOOLEAN:
+                    return (boolean)$value;
+                case self::TYPE_STRING:
+                    return (string)$value;
+                default:
+                    throw new InvalidParamException("Unsupported type '{$type}'");
+            }
```

Comment:
Would it be possible to support `DateTime` objects per default?
This is a bit tricky, because the format depends on the DB in wich it comes after find and if it comes from a form its based on a rule.

I would anyways recommend to give datetime fields from DBs more attention, because it is often used and always a pain.